src/StaticAssets/src/StaticAssetsInvoker.cs

```diff
@@ -0,0 +1,433 @@
+                    _length = long.Parse(length, CultureInfo.InvariantCulture);
+                    break;
+                default:
+                    _remainingHeaders ??= [];
+                    _remainingHeaders.Add(responseHeader);
+                    break;
+            }
+        }
+
+        if (_etag == null)
+        {
+            throw new InvalidOperationException("The ETag header is required.");
+        }
+    }
+
+    public string Route => _resource.Route;
+
+    public string PhysicalPath => _fileInfo.PhysicalPath ?? string.Empty;
+
+    private Task ApplyResponseHeadersAsync(StaticAssetInvocationContext context, int statusCode)
+    {
+        if (statusCode < 400)
+        {
+            // these headers are returned for 200, 206, and 304
+            // they are not returned for 412 and 416
+            if (!string.IsNullOrEmpty(_contentType))
+            {
+                context.Response.ContentType = _contentType;
+            }
+
+            var responseHeaders = context.ResponseHeaders;
+            responseHeaders.LastModified = _lastModified;
+            responseHeaders.ETag = _etag;
+            responseHeaders.Headers.AcceptRanges = "bytes";
+
+            foreach (var header in _remainingHeaders ?? [])
+            {
+                responseHeaders.Append(header.Name, header.Value);
+            }
+        }
+
+        return Task.CompletedTask;
+    }
+
+    private Task SendStatusAsync(StaticAssetInvocationContext context, int statusCode)
+    {
+        _logger.Handled(statusCode, Route);
+
+        // Only clobber the default status (e.g. in cases this a status code pages retry)
+        if (context.Response.StatusCode == StatusCodes.Status200OK)
+        {
+            context.Response.StatusCode = statusCode;
+        }
+
+        return ApplyResponseHeadersAsync(context, statusCode);
+    }
+
+    public async Task Invoke(HttpContext context)
+    {
+        var requestContext = new StaticAssetInvocationContext(
+            context,
+            _etag,
+            _lastModified,
+            _length,
+            _logger);
+
+        var (preconditionState, isRange, range) = requestContext.ComprehendRequestHeaders();
+        switch (preconditionState)
+        {
+            case PreconditionState.Unspecified:
+            case PreconditionState.ShouldProcess:
+                if (HttpMethods.IsHead(context.Request.Method))
+                {
+                    await SendStatusAsync(requestContext, StatusCodes.Status200OK);
+                    return;
+                }
+
+                try
+                {
+                    if (isRange)
+                    {
+                        await SendRangeAsync(requestContext, range);
+                        return;
+                    }
+                    else
+                    {
+                        context.Response.ContentLength = _length;
+                    }
+
+                    await SendAsync(requestContext);
```

Comment:
Nit: Removing the `else` block might make it clearer that the `ContentLength` gets unconditionally set before the call to `SendAsync` (because of the early return in the `if`). Alternatively, the code following the `else` block could be moved inside it. Again not a huge deal but I wasn't sure what the significance of the `else` block was when reviewing.